src/QsCompiler/Tests.Compiler/CompilationTrackerTests.fs

```diff
@@ -0,0 +1,320 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.Testing
+
+open System
+open System.Collections.Generic
+open System.IO
+open System.Reflection
+open System.Text.Json
+open System.Threading
+open Microsoft.FSharp.Core
+open Microsoft.Quantum.QsCompiler.CommandLineCompiler
+open Microsoft.Quantum.QsCompiler.Diagnostics
+open Xunit
+open Xunit.Abstractions
+
+type MeasureNestedTasksTestCases() as this =
+    inherit TheoryData<int[]>()
+    do this.Add([||])
+       this.Add([|1000;|])
+       this.Add([|150; 225|])
+       this.Add([|1000; 250; 825; 1555|])
+       this.Add([|250; 300; 400; 200; 150; 750; 425; 103; 900|])
+
+type MeasureDoubleNestedTasksTestCases() as this =
+    inherit TheoryData<int[], int[]>()
+    do this.Add([||], [||])
+       this.Add([|0|], [|500|])
+       this.Add([|500|], [|0;|])
+       this.Add([|500|], [|500|])
+       this.Add([|150; 325|], [|500; 250; 765|])
+       this.Add([|500; 250; 765|], [|150; 325|])
+       this.Add([|500;|], [|1035; 555; 775; 225; 2500; 15; 95|])
+       this.Add([|350; 155; 670; 2250; 25; 1110|], [|100|])
+       this.Add([|0;|], [|35; 625; 1275; 335; 1500; 155; 5|])
+       this.Add([|5; 5; 10; 5; 5|], [|35; 10; 5; 30; 15; 5|])
+
+type CompilationTrackerTests (output:ITestOutputHelper) =
+
+    let getResultsFolder (folderName: string) =
+        Path.Combine("Results", folderName)
+
+    let getResultsFile (folderName: string) =
+        let resultsFolder = getResultsFolder folderName
+        Path.Combine(resultsFolder, CompilationTracker.CompilationPerfDataFileName)
+
+    let getResultsDictionary (folderName: string) =
+        folderName |> getResultsFolder |> CompilationTracker.PublishResults
+        getResultsFile folderName |> File.ReadAllText |> JsonSerializer.Deserialize<IDictionary<string, int>>
+
+    [<Theory>]
+    [<InlineData(0)>]
+    [<InlineData(10)>]
+    [<InlineData(25)>]
+    [<InlineData(100)>]
+    [<InlineData(350)>]
+    [<InlineData(635)>]
+    [<InlineData(1000)>]
+    [<InlineData(3333)>]
+    [<InlineData(5050)>]
+    [<InlineData(10000)>]
+    member this.``Measure Task`` (durationInMs: int) =
+        CompilationTracker.ClearData()
+        let taskName = "TestTask"
+
+        // Measure time spent in a task.
+        CompilationTracker.OnCompilationTaskEvent(CompilationTaskEventType.Start, null, taskName)
+        Thread.Sleep durationInMs
+        CompilationTracker.OnCompilationTaskEvent(CompilationTaskEventType.End, null, taskName)
+
+        // Verify measured results are the expected ones.
+        let resultsDictionary = MethodBase.GetCurrentMethod().Name |> getResultsDictionary
+        let mutable measuredDurationInMs = 0
+        Assert.True(resultsDictionary.TryGetValue(taskName, &measuredDurationInMs))
+        Assert.InRange(measuredDurationInMs, durationInMs, Int32.MaxValue)
```

Comment:
It's unfortunate that a performance tracker that always reported `Int32.MaxValue` for all tasks would still pass the tests, but I'm not sure if it's worth trying to come up with a deterministic way to test the upper bound.

Reply:
Yes, that is unfortunate. I think we could consider adding a separate object (that implements a new interface) to perform time accounting such that in the tests a custom time accounting object can be implemented to yield the expected results.

I added this note to #726